xbmc/cores/VideoPlayer/Buffers/VideoBufferPoolDMA.cpp

```diff
@@ -0,0 +1,127 @@
+/*
+ *  Copyright (C) 2005-2020 Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#include "VideoBufferPoolDMA.h"
+
+#include "cores/VideoPlayer/Buffers/VideoBufferDMA.h"
+#include "threads/SingleLock.h"
+#include "utils/BufferObjectFactory.h"
+
+#include <drm_fourcc.h>
+
+extern "C"
+{
+#include <libavutil/pixfmt.h>
+}
+
+CVideoBufferPoolDMA::~CVideoBufferPoolDMA()
+{
+  CSingleLock lock(m_critSection);
+
+  for (auto buf : m_all)
+    delete buf;
+}
+
+CVideoBuffer* CVideoBufferPoolDMA::Get()
+{
+  CSingleLock lock(m_critSection);
+
+  CVideoBufferDMA* buf = nullptr;
+  if (!m_free.empty())
+  {
+    int idx = m_free.front();
+    m_free.pop_front();
+    m_used.push_back(idx);
+    buf = m_all[idx];
+  }
+  else
+  {
+    int id = m_all.size();
+    buf = new CVideoBufferDMA(*this, id, m_fourcc, m_size);
+
+    if (!buf->Alloc())
+    {
+      delete buf;
+      return nullptr;
+    }
+
+    m_all.push_back(buf);
+    m_used.push_back(id);
+  }
+
+  buf->Acquire(GetPtr());
+  return buf;
+}
+
+void CVideoBufferPoolDMA::Return(int id)
+{
+  CSingleLock lock(m_critSection);
+
+  m_all[id]->Unref();
+  auto it = m_used.begin();
+  while (it != m_used.end())
+  {
+    if (*it == id)
+    {
+      m_used.erase(it);
+      break;
+    }
+    else
+      ++it;
+  }
+  m_free.push_back(id);
+}
+
+void CVideoBufferPoolDMA::Configure(AVPixelFormat format, int size)
+{
+  CSingleLock lock(m_critSection);
+
+  m_fourcc = TranslateFormat(format);
+  m_size = static_cast<uint64_t>(size);
+}
+
+bool CVideoBufferPoolDMA::IsConfigured()
+{
+  CSingleLock lock(m_critSection);
+
+  return (m_fourcc != 0 && m_size != 0);
+}
+
+bool CVideoBufferPoolDMA::IsCompatible(AVPixelFormat format, int size)
+{
+  CSingleLock lock(m_critSection);
+
+  if (m_fourcc != TranslateFormat(format) || m_size != static_cast<uint64_t>(size))
+    return false;
+
+  return true;
+}
+
+void CVideoBufferPoolDMA::Released(CVideoBufferManager& videoBufferManager)
+{
+  if (!CBufferObjectFactory::CreateBufferObject(true))
+    return;
+
+  videoBufferManager.RegisterPool(std::make_shared<CVideoBufferPoolDMA>());
+}
+
+std::shared_ptr<IVideoBufferPool> CVideoBufferPoolDMA::CreatePool()
+{
+  return std::make_shared<CVideoBufferPoolDMA>();
+}
+
+uint32_t CVideoBufferPoolDMA::TranslateFormat(AVPixelFormat format)
+{
+  switch (format)
+  {
+    case AV_PIX_FMT_YUV420P:
```

Comment:
You can probably add `AV_PIX_FMT_YUVJ420P` as a known format, it should be treated same as `AV_PIX_FMT_YUV420P` with a `AVFrame.color_range=AVCOL_RANGE_JPEG`, to support it on the video codec side we need to do something like the following so that the `COLOR_RANGE` plane property get set correctly:
```
  pVideoPicture->color_range = m_pFrame->color_range == AVCOL_RANGE_JPEG ||
                               m_pFrame->format == AV_PIX_FMT_YUVJ420P ||
                               m_hints.colorRange == AVCOL_RANGE_JPEG;
```

```suggestion
    case AV_PIX_FMT_YUV420P:
    case AV_PIX_FMT_YUVJ420P:
```

Reply:
Hmm interesting, I've never come across this format before. FFmpeg defines it as:
```
planar YUV 4:2:0, 12bpp, full scale (JPEG), deprecated in favor of AV_PIX_FMT_YUV420P and setting color_range```

do you think it's worth adding a deprecated type?

Reply:
I have at least one sample H264 file captured with a GoPro that reported this format when ffmpeg h264 software decoding was used, this probably was with ffmpeg 4.0 so things may have changed (unlikely) with the update to 4.2.

Reply:
No worries, I'll add it in